### PR TITLE
feat: add rename and renamenx

### PR DESCRIPTION
**Pull request by zztaki**
#30
This PR aims to add rename and renamenx commands for pikiwidb.

Running `./pikiwidbtests.sh basic`
image

There are a few places that I think need to be focused on.

### CODE REVIEW

File: src/storage/src/storage.cc [hunk 1]
```diff
@@ -1754,6 +1754,59 @@ Status Storage::Keys(const DataType& data_type, const std::string& pattern, std:
...
+  // zzl TODO: Is Status::Corruption() appropriate?
+  count = Exists(newkeys);
+  if (count > 0) {
+    return Status::Corruption();
+  }
+
+  // zzl TODO: Is Status::Aborted() appropriate?
+  std::string value;
+  uint64_t ttl;
+  if (GetWithTTL(key, &value, &ttl) != Status::OK() || Del(keys) <= 0 || Set(newkey, value) != Status::OK()) {
```
Comment (zztaki) [inline src/storage/src/storage.cc:1800]: I'm not sure if calling other functions like GetWithTTL, Del, Set and Expire inside the rename(nx) function is allowed.

File: src/storage/src/storage.cc [hunk 2]
```diff
@@ -1754,6 +1754,59 @@ Status Storage::Keys(const DataType& data_type, const std::string& pattern, std:
...
+    return Status::Aborted();
```
Comment (zztaki) [inline src/storage/src/storage.cc:1801]: Is return Status::Aborted() reasonable?

File: tests/unit/basic.tcl [hunk 3]
```diff
@@ -291,89 +292,105 @@ start_server {tags {"basic"}} {
...
+
+    test {RENAMENX against non existing source key} {
+        catch {r renamenx nokey foobar} err
+        format $err
+    } {ERR*}
+
+    test {RENAME where source and dest key are the same (existing)} {
+        r set mykey foo
+        r rename mykey mykey
+    } {OK}
```
Comment (zztaki) [inline tests/unit/basic.tcl:363]: In fact, RENAME where source and dest key are the same (existing) is allowed. https://github.com/redis/redis/blame/f95031c4733078788063de775c968b6dc85522c0/tests/unit/keyspace.tcl#L134

File: src/storage/src/redis.h [hunk 4]
```diff
@@ -148,6 +148,18 @@ class Redis {
   virtual Status ZsetsTTL(const Slice& key, uint64_t* timestamp);
   virtual Status SetsTTL(const Slice& key, uint64_t* timestamp);
 
+  virtual Status StringsRename(const Slice& key, Redis* new_inst, const Slice& newkey);
```
Comment (zztaki) [inline src/storage/src/redis.h:158]: `key` and `newkey` may not be on the same shard. So I use `new_inst` to represent newkey's shard and pass it as an arg.

File: src/cmd_keys.cc [hunk 5]
```diff
@@ -295,4 +295,45 @@ void PttlCmd::DoCmd(PClient* client) {
...
+}
+
+void RenameNXCmd::DoCmd(PClient* client) {
+  storage::Status s =
+      PSTORE.GetBackend(client->GetCurrentDB())->GetStorage()->Renamenx(client->Key(), client->argv_[2]);
+  if (s.ok()) {
+    client->SetRes(CmdRes::kOK);
+  } else if (s.IsNotFound()) {
+    client->SetRes(CmdRes::kNotFound, s.ToString());
+  } else if (s.IsCorruption()) {
```
Comment (zztaki) [inline src/cmd_keys.cc:332]: I'm not sure if there is a better error code to represent that newkey already exists, thus returning (int) 0

File: src/cmd_keys.cc [hunk 6]
```diff
@@ -295,4 +295,45 @@ void PttlCmd::DoCmd(PClient* client) {
...
+    client->AppendInteger(0);  // newkey already exists
```
Comment (zztaki) [inline src/cmd_keys.cc:333]: I'm not sure if there is a better error code to represent that newkey already exists, thus returning `(int) 0`.

**Comment by dingxiaoshuai123** [comment]
May I ask if the command can be modified to include Go tests for validation?

**Comment by dingxiaoshuai123** [comment]
May I ask if it is possible to supplement some Go tests to validate the functionality of this command?

**Comment by zztaki** [comment]
> May I ask if it is possible to supplement some Go tests to validate the functionality of this command?

I haven't implemented it yet. Are there any good examples? :)

**Comment by dingxiaoshuai123** [comment]
> > May I ask if it is possible to supplement some Go tests to validate the functionality of this command?
> 
> I haven't implemented it yet. Are there any good examples? :)



> > May I ask if it is possible to supplement some Go tests to validate the functionality of this command?
> 
> I haven't implemented it yet. Are there any good examples? :)

There are Go test files under the path pikiwidb/tests, which can be referred to for inspiration.

**Comment by dingxiaoshuai123** [comment]
> > May I ask if it is possible to supplement some Go tests to validate the functionality of this command?
> 
> I haven't implemented it yet. Are there any good examples? :)

 https://github.com/redis/go-redis/blob/master/commands_test.go This might be helpful to you.

**Comment by zztaki** [comment]
> > > May I ask if it is possible to supplement some Go tests to validate the functionality of this command?
> > 
> > 
> > I haven't implemented it yet. Are there any good examples? :)
> 
> https://github.com/redis/go-redis/blob/master/commands_test.go This might be helpful to you.

Thank you for your help! I will complete it at night.

**Comment by zztaki** [comment]
cc

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
image

Current PR is blocked by this error.

**Comment by dingxiaoshuai123** [comment]
> image
> 
> Current PR is blocked by this error.

Please update the code.

**Comment by zztaki** [comment]
> > image
> > Current PR is blocked by this error.
> 
> Please update the code.

Done

**Review by Mixficsol — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**